rbac/api/cross_access/relation_api_dual_write_cross_access_handler.py

```diff
@@ -86,13 +91,58 @@ def _replicate(self):
                     },
                     partition_key=PartitionKey.byEnvironment(),
                     remove=self.relations_to_remove,
-                    add=self.relations_to_add,
+                    add=deduplicated_add,
                 ),
             )
         except Exception as e:
             logger.error("Error occurred in cross account replicate event", e)
             raise DualWriteException(e)
 
+    def _deduplicate_relations(self, relationships):
+        """
+        Deduplicate role_binding#subject relationships.
+
+        When generate_relations is called multiple times with the same user/role,
+        duplicate subject tuples are created. These are expected and should be deduplicated.
+        Other duplicates are bugs.
+        """
+        seen = set()
+        deduplicated = []
+        subject_duplicates = []
+
+        for rel in relationships:
+            key = stringify_spicedb_relationship(rel)
+
+            if key in seen:
+                # Check if this is a role_binding#subject tuple
+                is_subject_tuple = (
+                    rel.resource.type.namespace == "rbac"
+                    and rel.resource.type.name == "role_binding"
+                    and rel.relation == "subject"
+                )
+
+                if is_subject_tuple:
+                    # Expected duplicate - generate_relations called multiple times
+                    subject_duplicates.append(key)
+                    continue
+                else:
+                    # Unexpected duplicate - this is a bug!
+                    raise ValueError(
+                        f"Unexpected duplicate relationship in CAR handler: {key}. "
+                        "This indicates a bug in tuple generation logic."
+                    )
+
+            seen.add(key)
+            deduplicated.append(rel)
+
+        if subject_duplicates:
+            logger.info(
+                f"[CAR Dual Write] Deduplicated {len(subject_duplicates)} subject tuples "
+                f"(had {len(relationships)}, kept {len(deduplicated)})"
+            )
+
+        return deduplicated
```

Comment:
We could probably make this a reusable method in the superclass, since we need it in both places.